In [1]:
import warnings
warnings.filterwarnings('ignore')
import pandas as pd
import numpy as np

# Display and Plotting
import matplotlib.pylab as plt
import seaborn as sns

# seaborn plotting style
sns.set(style='ticks', context='poster')

In [2]:
filename_ts = r'C:\Users\60547\Documents\py\python_lianxi\datasets\datawhile\data.csv'
df=pd.read_csv(filename_ts,encoding='gbk')
print(df.columns)
df.head(5)

Index(['unnamedid', 'custid', 'trade_no', 'bank_card_no', 'low_volume_percent',
       'middle_volume_percent', 'take_amount_in_later_12_month_highest',
       'trans_amount_increase_rate_lately', 'trans_activity_month',
       'trans_activity_day', 'transd_mcc', 'trans_days_interval_filter',
       'trans_days_interval', 'regional_mobility', 'student_feature',
       'repayment_capability', 'is_high_user', 'number_of_trans_from_2011',
       'first_transaction_time', 'historical_trans_amount',
       'historical_trans_day', 'rank_trad_1_month', 'trans_amount_3_month',
       'avg_consume_less_12_valid_month', 'abs',
       'top_trans_count_last_1_month', 'avg_price_last_12_month',
       'avg_price_top_last_12_valid_month', 'reg_preference_for_trad',
       'trans_top_time_last_1_month', 'trans_top_time_last_6_month',
       'consume_top_time_last_1_month', 'consume_top_time_last_6_month',
       'cross_consume_count_last_1_month',
       'trans_fail_top_count_enum_last_1_month',
    

unnamedid   custid      trade_no bank_card_no  low_volume_percent  \
0          5  2791858  2.018050e+31          卡号1                0.01   
1         10   534047  2.018050e+31          卡号1                0.02   
2         12  2849787  2.018050e+31          卡号1                0.04   
3         13  1809708  2.018050e+31          卡号1                0.00   
4         14  2499829  2.018050e+31          卡号1                0.01   

   middle_volume_percent  take_amount_in_later_12_month_highest  \
0                   0.99                                      0   
1                   0.94                                   2000   
2                   0.96                                      0   
3                   0.96                                   2000   
4                   0.99                                      0   

   trans_amount_increase_rate_lately  trans_activity_month  \
0                               0.90                  0.55   
1                               1.28                  1.00   
2                               1.00                  1.00   
3                               0.13                  0.57   
4                               0.46                  1.00   

   trans_activity_day        ...         loans_max_limit  loans_avg_limit  \
0               0.313        ...                  2900.0           1688.0   
1               0.458        ...                  3500.0           1758.0   
2               0.114        ...                  1600.0           1250.0   
3               0.777        ...                  3200.0           1541.0   
4               0.175        ...                  2300.0           1630.0   

   consfin_credit_limit  consfin_credibility  consfin_org_count_current  \
0                1200.0                 75.0                        1.0   
1               15100.0                 80.0                        5.0   
2                4200.0                 87.0                        1.0   
3               16300.0                 80.0                        5.0   
4                8300.0                 79.0                        2.0   

   consfin_product_count  consfin_max_limit  consfin_avg_limit  \
0                    2.0             1200.0             1200.0   
1                    6.0            22800.0             9360.0   
2                    1.0             4200.0             4200.0   
3                    5.0            30000.0            12180.0   
4                    2.0             8400.0             8250.0   

   latest_query_day  loans_latest_day  
0              12.0              18.0  
1               4.0               2.0  
2               2.0               6.0  
3               2.0               4.0  
4              22.0             120.0  

[5 rows x 90 columns]

In [3]:
#查看object类型的数据  值的分布
print(df.bank_card_no.value_counts(),'\n')
print(df.reg_preference_for_trad.value_counts(),'\n')
print(df.source.value_counts(),'\n')
print(df.student_feature.value_counts())
df.select_dtypes(include=['object']).head(5)


卡号1    4754
Name: bank_card_no, dtype: int64 

一线城市    3403
三线城市    1064
境外       150
二线城市     131
其他城市       4
Name: reg_preference_for_trad, dtype: int64 

xs    4754
Name: source, dtype: int64 

1.0    1754
2.0       2
Name: student_feature, dtype: int64


bank_card_no reg_preference_for_trad source id_name latest_query_time  \
0          卡号1                    一线城市     xs      蒋红         2018/4/25   
1          卡号1                    一线城市     xs     崔向朝          2018/5/3   
2          卡号1                    一线城市     xs     王中云          2018/5/5   
3          卡号1                    三线城市     xs     何洋洋          2018/5/5   
4          卡号1                    一线城市     xs      赵洋         2018/4/15   

  loans_latest_time  
0         2018/4/19  
1          2018/5/5  
2          2018/5/1  
3          2018/5/3  
4          2018/1/7

In [4]:
#student_feature缺失太多直接丢弃,其他没有业务价值
drop_columns=['bank_card_no','source','id_name','unnamedid','custid','student_feature','trade_no']
df.drop(drop_columns,axis=1,inplace=True)

#日期处理为日期格式datetime
df['latest_query_time']=pd.to_datetime(df['latest_query_time'])
df['loans_latest_time']=pd.to_datetime(df['loans_latest_time'])
df.select_dtypes(include=['object']).head(5)

#将城市进行数值化
map_dict={'一线城市':1,'二线城市':2,'三线城市':3,'境外':4,'其他':0}
df['reg_preference_for_trad']=df['reg_preference_for_trad'].map(map_dict)
df.head()

low_volume_percent  middle_volume_percent  \
0                0.01                   0.99   
1                0.02                   0.94   
2                0.04                   0.96   
3                0.00                   0.96   
4                0.01                   0.99   

   take_amount_in_later_12_month_highest  trans_amount_increase_rate_lately  \
0                                      0                               0.90   
1                                   2000                               1.28   
2                                      0                               1.00   
3                                   2000                               0.13   
4                                      0                               0.46   

   trans_activity_month  trans_activity_day  transd_mcc  \
0                  0.55               0.313        17.0   
1                  1.00               0.458        19.0   
2                  1.00               0.114        13.0   
3                  0.57               0.777        22.0   
4                  1.00               0.175        13.0   

   trans_days_interval_filter  trans_days_interval  regional_mobility  \
0                        27.0                 26.0                3.0   
1                        30.0                 14.0                4.0   
2                        68.0                 22.0                1.0   
3                        14.0                  6.0                3.0   
4                        66.0                 42.0                1.0   

         ...         loans_max_limit  loans_avg_limit  consfin_credit_limit  \
0        ...                  2900.0           1688.0                1200.0   
1        ...                  3500.0           1758.0               15100.0   
2        ...                  1600.0           1250.0                4200.0   
3        ...                  3200.0           1541.0               16300.0   
4        ...                  2300.0           1630.0                8300.0   

   consfin_credibility  consfin_org_count_current  consfin_product_count  \
0                 75.0                        1.0                    2.0   
1                 80.0                        5.0                    6.0   
2                 87.0                        1.0                    1.0   
3                 80.0                        5.0                    5.0   
4                 79.0                        2.0                    2.0   

   consfin_max_limit  consfin_avg_limit  latest_query_day  loans_latest_day  
0             1200.0             1200.0              12.0              18.0  
1            22800.0             9360.0               4.0               2.0  
2             4200.0             4200.0               2.0               6.0  
3            30000.0            12180.0               2.0               4.0  
4             8400.0             8250.0              22.0             120.0  

[5 rows x 83 columns]

In [ ]:
#统计缺失值
print('before:',df.isnull().sum())
#数量较小，直接丢弃
df.dropna(inplace=True)
print('number after processing:',df.shape)

before: low_volume_percent                          2
middle_volume_percent                       2
take_amount_in_later_12_month_highest       0
trans_amount_increase_rate_lately           3
trans_activity_month                        2
trans_activity_day                          2
transd_mcc                                  2
trans_days_interval_filter                  8
trans_days_interval                         2
regional_mobility                           2
repayment_capability                        0
is_high_user                                0
number_of_trans_from_2011                   2
first_transaction_time                      2
historical_trans_amount                     0
historical_trans_day                        2
rank_trad_1_month                           2
trans_amount_3_month                        0
avg_consume_less_12_valid_month             2
abs                                         0
top_trans_count_last_1_month                2
avg_price_last_12_month   

In [ ]:
#两列日期格式数据处理直接丢弃
datatime_data=['loans_latest_time','latest_query_time']
df.drop(datatime_data,axis=1,inplace=True)
#丢掉方差为0即没有变化的列
df.drop(df.columns[df.std()==0],axis=1,inplace=True)
df.shape

(3979, 81)

In [ ]:
#热力图研究相关性
plt.figure(figsize=(62,54))
corr=df.drop(['status'],axis=1).corr()
sns.heatmap(corr,fmt='.1g',annot=True,cbar=False)


In [ ]:
#将数据进行标准化
from sklearn.preprocessing import StandardScaler
train_x=df.drop(['status'],axis=1)
train_y=df['status']
train_x=StandardScaler().fit_transform(train_x)
print(train_x,'\n',train_x.shape)

In [ ]:
#拆分测试集和训练集
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(train_x,train_y,test_size=0.3,random_state=2018)
print(x_train.shape)
print(x_test.shape)
print(y_test.shape)

In [ ]:
#引入模型进行训练
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score


In [ ]:
#引入混淆矩阵对分类的结果进行可视化
from sklearn.metrics import confusion_matrix
import itertools
def plot_confusion_matrix(cm, classes,
                          title='Confusion matrix',
                          cmap=plt.cm.Blues):
    """
    This function prints and plots the confusion matrix.
    """
    plt.imshow(cm, interpolation='nearest', cmap=cmap)
    plt.title(title)
    plt.colorbar()
    tick_marks = np.arange(len(classes))
    plt.xticks(tick_marks, classes, rotation=0)
    plt.yticks(tick_marks, classes)

    thresh = cm.max() / 2.
    for i, j in itertools.product(range(cm.shape[0]), range(cm.shape[1])):
        plt.text(j, i, cm[i, j],
                 horizontalalignment="center",
                 color="white" if cm[i, j] > thresh else "black")

    plt.tight_layout()
    plt.ylabel('True label')
    plt.xlabel('Predicted label')

In [ ]:
knn=KNeighborsClassifier(n_neighbors=5)
knn.fit(x_train,y_train)

y_pred=knn.predict(x_test)
knn.score(x_test,y_test)

In [ ]:
print('recall score of randomforest is:%.3f'%recall_score(y_test,y_pred))
print('precision score of randomforest is:%.3f'%precision_score(y_test,y_pred))  

In [ ]:
cnf_knn=confusion_matrix(y_test,y_pred)
class_names=[0,1]
plot_confusion_matrix(cnf_knn,class_names)

#效果较差，尝试用更复杂的集成算法随机森林机型训练

In [ ]:
rf=RandomForestClassifier()
rf.fit(x_train,y_train)
y_pred=rf.predict(x_test)
rf.score(x_test,y_test)


In [ ]:
print('recall score of randomforest is:%.3f'%recall_score(y_test,y_pred))
print('precision score of randomforest is:%.3f'%precision_score(y_test,y_pred))  

In [ ]:
cnf_rf=confusion_matrix(y_test,y_pred)
plot_confusion_matrix(cnf_rf,class_names)

总结:选取两个模型KNN和随机森林，丢弃了9列，保留81列特征列分析(包含标签列status)
两个模型总体分类的正确率接近80%，随机森林较高(77.8%),查准率也能达到64%(随机森林,knn为55%),但查全率均较低(27%，说明很多status=1的类被没有被正确分类),下一步的动作:对特征列机型进一步筛选(目前列太多，关系热力图看不出高线性相关列)